singularity-eos/base/fast-math/logs.hpp

```diff
@@ -78,76 +62,166 @@
+double lg_o2_portable(const double x) {
+  PORTABLE_REQUIRE(x > 0, "log divergent for x <= 0");
+  constexpr double four_thirds = 4. / 3.;
+  int e;
+  const double m = frexp(x, &e);
+  return e - four_thirds * (m - 2) * (m - 1);
+}
+
+// This version uses the exact formula
+PORTABLE_FORCEINLINE_FUNCTION
+double pow2_o2_portable(const double x) {
+  // log2(mantissa). should go between -1 and 0
+  const int flr = std::floor(x);
+  const double lm = x - flr - 1;
+  const double mantissa = 0.5 * (3 - std::sqrt(1 - 3 * lm));
+  const double exponent = flr + 1;
+  return ldexp(mantissa, exponent);
+}
+
+// Integer aliased/bithacked versions
+PORTABLE_FORCEINLINE_FUNCTION
+double lg_o2_aliased(const double x) {
+  PORTABLE_REQUIRE(x > 0, "log divergent for x <= 0");
+  // as_int(1.0) == 2^62 - 2^52
+  constexpr std::int64_t one_as_int = 4607182418800017408;
+  // 1/(as_int(2.0) - as_int(1.0)) == 2^-52
+  constexpr double scale_down = 2.220446049250313e-16;
+  // 2^52 - 1
+  constexpr std::int64_t mantissa_mask = 4503599627370495;
+  // 2^26 - 1
+  constexpr std::int64_t low_mask = 67108863;
+
+  const std::int64_t x_as_int = as_int(x) - one_as_int;
+  const std::int64_t frac_as_int = x_as_int & mantissa_mask;
+  const std::int64_t frac_high = frac_as_int >> 26;
+  const std::int64_t frac_low = frac_as_int & low_mask;
+  const std::int64_t frac_squared =
+      frac_high * frac_high + ((frac_high * frac_low) >> 25);
+
+  return static_cast<double>(x_as_int + ((frac_as_int - frac_squared) / 3)) * scale_down;
+}
+
+PORTABLE_FORCEINLINE_FUNCTION
+double pow2_o2_aliased(const double x) {
+  // as_int(1.0) == 2^62 - 2^52
+  constexpr std::int64_t one_as_int = 4607182418800017408;
+  // as_int(2.0) - as_int(1.0) == 2^52
   constexpr double scale_up = 4503599627370496;
-  return as_double(static_cast<long long int>(x * scale_up) + one_as_long_int);
+  constexpr std::int64_t mantissa_mask = 4503599627370495; // 2^52 - 1
+  constexpr std::int64_t a = 9007199254740992;             // 2 * 2^52
+  constexpr double b = 67108864;                           // 2^26
+  constexpr std::int64_t c = 18014398509481984;            // 4 * 2^52
+
+  const std::int64_t x_as_int = static_cast<std::int64_t>(x * scale_up);
+  const std::int64_t frac_as_int = x_as_int & mantissa_mask;
+  const std::int64_t frac_sqrt =
+      static_cast<std::int64_t>(b * std::sqrt(static_cast<double>(c - 3 * frac_as_int)));
+
+  return as_double(x_as_int + a - frac_sqrt - frac_as_int + one_as_int);
 }
+// ----------------------------------------------------------------------
 
 PORTABLE_FORCEINLINE_FUNCTION
 double fastlg(const double x) {
-#ifdef SINGULARITY_USE_HIGH_RISK_MATH
-  return fastlg_aliased(x);
+#ifdef SINGULARITY_NQT_ORDER_1
+#ifdef SINGULARITY_NQT_PORTABLE
+  return lg_o1_portable(x);
 #else
-  int n;
-  assert(x > 0 && "log divergent for x <= 0");
-#ifndef SINGULARITY_USE_SINGLE_LOGS
-  // double precision is default
-  const double y = frexp(x, &n);
+  return lg_o1_aliased(x);
```

Comment:
Since we're already here, can we move to `if constexpr`? Note that we'd probably want to translate preproccessor defines into some kind of typed boolean or enum values. You're call.

Reply:
Agree that would be more readable. Let me play with it.

Reply:
The proper way to do this would be to get `CMake` to emit booleans with generated code, but I got lazy and stuck `constexpr bool`s in a `Settings` namespace.